### PR TITLE
Add metrics for cache entry memory size

### PR DESCRIPTION
**Pull request by BrynCooke**
Query planner cache entries may use an significant amount of memory in the Router.
To help users understand and monitor this the Router now exposes a new metric `apollo.router.cache.storage.estimated_size`.

This metric give an estimated size in bytes for the cache entry and has the following attributes:
- `kind`: `query planner`.
- `storage`: `memory`.

As the size is only an estimation, users should check for correlation with pod memory usage to determine if cache needs to be updated.

Usage scenario:
1. Your pods are being terminated due to memory pressure.
2. Add the following metrics to your monitoring system to track:
  * `apollo.router.cache.storage.estimated_size`.
  * `apollo_router_cache_size`.
  * ratio of `apollo_router_cache_hits` - `apollo_router_cache_misses`.
3. Observe the `apollo.router.cache.storage.estimated_size` to see if it grows over time and correlates with pod memory usage.
4. Observe the ratio of cache hits to misses to determine if the cache is being effective.

Remediation:
* Adjust the cache size to lower if the cache reaches near 100% hit rate but the cache size is still growing.
* Increase the pod memory to higher if cache hit rate is low and cache size is still growing.
* Adjust the cache size to lower if the latency of query planning cache misses is acceptable and memory availability is limited.

## Technical info
The estimate is only implemented for query plans and uses serde to create the estimate without actually writing to a string. It is very difficult to provide accurate memory usage, so it's only useful to see a trend.

Providing an estimated size for a cache entry is optional, and only implemented for query planner at this time.
Where no estimated size is set the gauge will not be emitted, e.g. APQ or entity caching.

## Testing
For manual testing we will deploy internally and observe that the metric tops out when the query cache is full.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: apollo-router/src/cache/size_estimation.rs
Comment (IvanGoncharov) [inline apollo-router/src/cache/size_estimation.rs:136]: Just to double check: `None` would still use the size of `Some` in memory, right?
We estimate it as 0 because of serve limitations?

File: apollo-router/src/cache/storage.rs
Comment (Geal) [inline apollo-router/src/cache/storage.rs:270]: the in memory cache lock is held until the end of the function, it should only be held when inserting the old entry.
Maybe fixed with something like:
```rust
let old_entry = {
  self.inner.lock().await.push(key.clone(), value.clone())
};
if let Some((_, v)) = old_entry {
```

**Review by Geal — CHANGES_REQUESTED** [review]

File: apollo-router/src/cache/storage.rs
Comment (BrynCooke) [reply apollo-router/src/cache/storage.rs]: Fixed in: [e37c87e](https://github.com/apollographql/router/pull/5770/commits/e37c87ebc97cb588dfc1de831760df1d0ce1d68f)

File: apollo-router/src/cache/size_estimation.rs
Comment (BrynCooke) [reply apollo-router/src/cache/size_estimation.rs:136]: Yeah, I guess in theory we can do something better. But None values won't have strings etc in them so I don't think we should worry about this and maybe we can follow up later to improve things.

**Review by Geal — APPROVED** [review]

File: .changesets/feat_enhanced_observability.md
Comment (BrynCooke) [inline .changesets/feat_enhanced_observability.md:12]: @shorgi Maybe this info should also be in the docs under cache troubleshooting?

**Comment by BrynCooke** [comment]
Going to wait for a docs review before merging.

File: .changesets/feat_enhanced_observability.md
Comment (shorgi) [reply .changesets/feat_enhanced_observability.md:12]: This is great content for docs. Since it's related to pod memory pressure, moving it to the k8s page.

Comment (shorgi) [inline .changesets/feat_enhanced_observability.md:12]:
```suggestion
```

File: .changesets/feat_enhanced_observability.md
Comment (shorgi) [inline .changesets/feat_enhanced_observability.md:25]:
```suggestion
To learn how to troubleshoot with this metric, see the [Pods terminating due to memory pressure](https://www.apollographql.com/docs/router/containerization/kubernetes#pods-terminating-due-to-memory-pressure) guide in docs.
```

**Review by shorgi — COMMENTED** [review]
Moved the usage scenario from the changeset into docs

**Comment by BrynCooke** [comment]
ROUTER-669

**Pull request merged**
